[JSC] Add V8 style ICU DST offset cache

Constellation opened the pull request:
#### 1f84372b0336751caf1a2ae482a1331e754e6b51
<pre>
[JSC] Add V8 style ICU DST offset cache
<a href="https://bugs.webkit.org/show_bug.cgi?id=258186">https://bugs.webkit.org/show_bug.cgi?id=258186</a>
rdar://110879575

Reviewed by Mark Lam.

This patch adds V8 style ICU DST offset cache into our DateCache.
Previously, we had one-way cache. But in some benchmarks, we observed
that frequent fallback to ICU due to cache miss. So we extend the cache
to 32-way LRU, which is what V8 is doing. So long as we hit this cache,
we do not need to query to ICU about DST offset.

* Source/JavaScriptCore/runtime/JSDateMath.cpp:
(JSC::DateCache::DSTCache::leastRecentlyUsed):
(JSC::DateCache::DSTCache::extendTheAfterCache):
(JSC::DateCache::DSTCache::localTimeOffset):
(JSC::DateCache::resetIfNecessarySlow):
(JSC::DateCache::localTimeOffset): Deleted.
* Source/JavaScriptCore/runtime/JSDateMath.h:
(JSC::LocalTimeOffsetCache::isEmpty):
(JSC::DateCache::DSTCache::DSTCache):
(JSC::DateCache::DSTCache::bumpEpoch):
(JSC::DateCache::DSTCache::reset):
(JSC::DateCache::localTimeOffset):

Canonical link: <a href="https://commits.webkit.org/265258@main">https://commits.webkit.org/265258@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b313b10f4d6d24cf341ced07a0d11121bd8cf703

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10413 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10645 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10919 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/12054 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9989 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/12748 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10597 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/12957 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/10571 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/11361 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/8771 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12452 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/8665 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/9428 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/16695 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/8792 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/9756 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/9580 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/12835 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/9869 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10013 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/8121 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/10542 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9171 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/2746 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13426 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/10831 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1164 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9872 "Built successfully") | | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/3/builds/2676 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/JavaScriptCore/runtime/JSDateMath.cpp:
```diff
@@ -185,103 +189,181 @@ LocalTimeOffset DateCache::calculateLocalTimeOffset(double millisecondsFromEpoch
+
+LocalTimeOffset DateCache::DSTCache::localTimeOffset(DateCache& dateCache, int64_t millisecondsFromEpoch, WTF::TimeType inputTimeType)
+{
+    if (millisecondsFromEpoch >= -WTF::Int64Milliseconds::maxECMAScriptTime && millisecondsFromEpoch <= WTF::Int64Milliseconds::maxECMAScriptTime) {
```
nit: Maybe this should be `millisecondsFromEpoch >= -minECMAScriptTime` since we have `minECMAScriptTime` defined as -8.64E15 (i.e. -maxECMAScriptTime) in JSDateMath.h.

Should the definition of `minECMAScriptTime` should be changed to be `-WTF::maxECMAScriptTime` instead of using the literal?  I'm not sure which is best.

OTOH, v8 uses 0 for minECMAScriptTime instead.  Do you know why they don't use `-WTF::maxECMAScriptTime`?

Constellation replied:
> OTOH, v8 uses 0 for minECMAScriptTime instead. Do you know why they don't use -WTF::maxECMAScriptTime?

V8 is conservative here to support non ICU TimeZones (so avoiding negative numbers). But we do not need to do it.
And rather, we should support negative one since it makes Gregorian / Julius calendar difference unhandled.

Constellation replied:
I defined `WTF:: Int64Milliseconds ::minECMAScriptTime`.

MenloDorian replied:
Sounds good.

MenloDorian reviewed (approved):
r=me with suggestion and question.

Constellation committed:
[JSC] Add V8 style ICU DST offset cache
https://bugs.webkit.org/show_bug.cgi?id=258186
rdar://110879575

Reviewed by Mark Lam.

This patch adds V8 style ICU DST offset cache into our DateCache.
Previously, we had one-way cache. But in some benchmarks, we observed
that frequent fallback to ICU due to cache miss. So we extend the cache
to 32-way LRU, which is what V8 is doing. So long as we hit this cache,
we do not need to query to ICU about DST offset.

* Source/JavaScriptCore/runtime/JSDateMath.cpp:
(JSC::DateCache::DSTCache::leastRecentlyUsed):
(JSC::DateCache::DSTCache::extendTheAfterCache):
(JSC::DateCache::DSTCache::localTimeOffset):
(JSC::DateCache::resetIfNecessarySlow):
(JSC::DateCache::localTimeOffset): Deleted.
* Source/JavaScriptCore/runtime/JSDateMath.h:
(JSC::LocalTimeOffsetCache::isEmpty):
(JSC::DateCache::DSTCache::DSTCache):
(JSC::DateCache::DSTCache::bumpEpoch):
(JSC::DateCache::DSTCache::reset):
(JSC::DateCache::localTimeOffset):

Canonical link: https://commits.webkit.org/265258@main
